core/src/main/scala/io/iteratee/Enumerator.scala

```diff
@@ -222,23 +246,124 @@ final object Enumerator extends EnumeratorInstances {
    */
   final def iterateUntilM[F[_], E](init: E)(f: E => F[Option[E]])(implicit F: Monad[F]): Enumerator[F, E] =
     new Enumerator[F, E] {
-      private[this] def loop[A](step: Step[F, E, A], last: Option[E]): F[Step[F, E, A]] =
-        last match {
-          case Some(last) if !step.isDone => F.flatMap(step.feedEl(last))(next => F.flatMap(f(last))(loop(next, _)))
-          case _ => F.pure(step)
-        }
-      final def apply[A](s: Step[F, E, A]): F[Step[F, E, A]] = loop(s, Some(init))
+      final def apply[A](s: Step[F, E, A]): F[Step[F, E, A]] = F.tailRecM((s, Some(init): Option[E])) {
+        case (step, Some(last)) if !step.isDone =>
+          F.map2(step.feedEl(last), f(last))((s1, next) => Left((s1, next)))
+        case (step, _) => F.pure(Right(step))
+      }
     }
 
   /**
    * An enumerator that returns the result of an effectful operation until
    * `None` is generated.
    */
-  final def generateM[F[_], E](f: F[Option[E]])(implicit F: Monad[F]): Enumerator[F, E] = new Enumerator[F, E] {
-    final def apply[A](s: Step[F, E, A]): F[Step[F, E, A]] =
-      if (s.isDone) F.pure(s) else F.flatMap(f) {
-        case None => F.pure(s)
-        case Some(e) => F.flatMap(s.feedEl(e))(apply)
+  final def generateM[F[_], E](f: F[Option[E]])(implicit F: Monad[F]): Enumerator[F, E] =
+    new Enumerator[F, E] {
+      final def apply[A](s: Step[F, E, A]): F[Step[F, E, A]] = F.tailRecM(s) { step =>
+        if (step.isDone) F.pure(Right(step)) else F.flatMap(f) {
+          case Some(e) => F.map(step.feedEl(e))(Left(_))
+          case None => F.pure(Right(step))
+        }
+      }
+    }
+
+  /**
+   * Enumerators that rely on `F` to provide stack safety.
+   *
+   * These implementations are more efficient than the default ones, but will
```

Comment:
Hmm, I meant the comparison to apply in the case where `F`'s `flatMap` is already stack safe, in which case it would be pretty strange for `Enumerator.iterate` not to have overhead `Enumerator.StackUnsafe.iterate` doesn't. I've updated the language, though.